### DISCUSSION TITLE
Volume integration of a density

### DISCUSSION THREAD

**Discussion by junyunPE** [MPh-py/MPh#123] [General]
Hi MPh，

I am trying to get the volume integration of magnetic density for a component using MPh. The realization is easily realized in comsol as shown in the picture, but I am not sure how to obtain it by MPh.
![picture](https://user-images.githubusercontent.com/101173566/221785633-c525d62c-9f4d-4fbe-82f0-09930c9e355f.png)

**Comment by john-hen** [comment]
General speaking, when you do something in the GUI, you can also do it in code, see #86. However, in this particular case I would not recommend doing it that way as you'll end up with much more code than is needed.

While MPh does not support evaluating "Volume Integration" features (even though it could, in principle, see #26), you can achieve the same by defining an "Integration" nonlocal coupling operator. Then it is a "global" evaluation, which [`model.evaluate()`] does support.

[`model.evaluate()`]: https://mph.readthedocs.io/en/1.2/api/mph.Model.html#mph.Model.evaluate

**Reply by junyunPE** [reply]
Thank you for your reply. I try to use the nonlocal coupling operator to realize my purpose in Comsol firstly. But I have some problems. It seems this expression is only valid for comp1, whose definition is air. However, it doesn't work for other components.
![Uploading picture.png…]()

**Reply by junyunPE** [reply]
![picture](https://user-images.githubusercontent.com/101173566/221814293-4e287242-488d-4cab-a539-0b0af2fa4818.png)

**Reply by john-hen** [reply]
I guess you mean that the "selection" is then part of the definition of the operator, and that you've selected "air". That is true. You can either define multiple operators, one for every selection you're interested in. Or you find out how to "change the selection" of the operator, then "update the solution" in code. (Again, refer to ["Access the full Comsol API"](https://mph.readthedocs.io/en/1.2/demonstrations.html#access-the-full-comsol-api) on how to do that.) I'd probably do the first thing, as that's straightforward and will run faster. Just means you'll have more operators defined in your model.
